Sources/Account/Account.swift

```diff
@@ -46,12 +46,28 @@ final class Account {
         return knowableBlockCount
     }
 
+    @available(*, deprecated, message: "Use cachedBalance(for:TokenId)")
     var cachedBalance: Balance {
+        cachedBalance(for: .MOB)
+    }
+
+    var cachedTxOutTokenIds: Set<TokenId> {
+        Set(allTxOutTrackers
+            .map { $0.knownTxOut.tokenId })
+    }
+
+    func cachedBalance(for tokenId: TokenId) -> Balance {
         let blockCount = knowableBlockCount
         let txOutValues = allTxOutTrackers
             .filter { $0.receivedAndUnspent(asOfBlockCount: blockCount) }
+            .filter { $0.knownTxOut.tokenId == tokenId }
             .map { $0.knownTxOut.value }
-        return Balance(values: txOutValues, blockCount: blockCount)
+        return Balance(values: txOutValues, blockCount: blockCount, tokenId: tokenId)
+    }
+
+    var cachedBalances: Balances {
+        let balances = cachedTxOutTokenIds.map { cachedBalance(for: $0) }
```

Comment:
I like the elegance of this, but it's also going to scan all the txOuts completely just to get the set of tokenIDs, and then per tokenID will do full passes over the txOuts again to check for received and unspent, and another for the tokenID check...

It can all be done with one pass over the txOuts... 

Right now not really a big deal, but if we have a lot of tokenIDs and if there is a lot of activity, could this be a potential performance problem?

Reply:
again, not sure the changes make sense here, happy to be provided evidence that it is significant, in the meantime we could add some new cases to the PerformanceTests to benchmark these functions but Id like to punt that to future work in the interest of the timelines we are on.

Reply:
performances tests added, the filtering is not the bottleneck, its the creation of the `Balance` structs, and im not sure optimizing that should be in the scope of this PR especially when we know for the near future there will only be at max 2 `Balance` structs.

Reply:
performance slows down at > 1000 balance structs.